### PR TITLE
Refactored PyTorch Autonencoder built up (Proposal to solve #434)

### LINKED ISSUE #434: Pytorch Autoencoder starts with Batchnorm before first Layer
Dear Contributors,

if BatchNorm is enabled in the options the PyTorch AutoEncoder starts with a BatchNorm before passing the input samples to the first linear layer. This might be unwanted behavior. Normally the input data should be preprocessed and passed into the first linear layer.

IMHO, BatchNorm should happen after the data has been processed by the first layer. Having preprocessing by Normalization and immediately applying BatchNorm to the data would be somehow doubled.

This behavior would then also differ from the TensorFlow implementation where the data gets passed directly into the first linear layer:
https://github.com/yzhao062/pyod/blob/master/pyod/models/auto_encoder.py#L172-L179

EDIT: I just noticed there is also a dropout layer after the last linear layer (and activation). This should IMHO also not be the case, as this would purposely set elements of the reconstructed sample (output) to zero. This also differs from the TensorFlow implementation.
See https://github.com/yzhao062/pyod/blob/master/pyod/models/auto_encoder.py#L190-L191
Also the DropOut is applied before the batch norm which will definitely skew the result. The batch norm won't be able to learn a good normalization if dropout is randomly applied right before (and turned off for inference).

**TL;DR**
--------------------------------------------------------------------------------------------------
Currently the order is: Input->BatchNorm->Neurons->Activation->DropOut->BatchNorm->....
IMHO it should be:     Input->Neurons->BatchNorm->Activation->DropOut->Neurons->....

--------------------------------------------------------------------------------------------------
I verified my suspicion by running the [PyTorch AutoEncoder Example](https://github.com/yzhao062/pyod/blob/master/examples/auto_encoder_torch_example.py) with `print(cls)` after Line 39.

The output was the following:
<Details>

```
AutoEncoder(batch_norm=True, batch_size=32, contamination=0.1,
      device=device(type='cpu'), dropout_rate=0.2, epochs=10,
      hidden_activation='relu', hidden_neurons=[64, 32],
      learning_rate=0.001, loss_fn=MSELoss(), preprocessing=True,
      weight_decay=1e-05)
inner_autoencoder(
  (activation): ReLU()
  (encoder): Sequential(
    (batch_norm0): BatchNorm1d(300, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (linear0): Linear(in_features=300, out_features=64, bias=True)
    (relu0): ReLU()
    (dropout0): Dropout(p=0.2, inplace=False)
    (batch_norm1): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (linear1): Linear(in_features=64, out_features=32, bias=True)
    (relu1): ReLU()
    (dropout1): Dropout(p=0.2, inplace=False)
  )
  (decoder): Sequential(
    (batch_norm0): BatchNorm1d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (linear0): Linear(in_features=32, out_features=64, bias=True)
    (dropout0): Dropout(p=0.2, inplace=False)
    (batch_norm1): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (linear1): Linear(in_features=64, out_features=300, bias=True)
    (dropout1): Dropout(p=0.2, inplace=False)
  )
)
```

</Details>

There you can see that the first element in the sequential of the encoder is a BatchNorm. A Pullrequest resolving this issue is on the way. Thanks for reading! (EDIT: See Pullrequest  #435 )

If you are annoyed by my fine-grained Pullrequests or issues, please let me know. I don't want to cause any extra work.

### PR DESCRIPTION
**Pull request by Lucew**
This Pullrequest is a proposal to fix #434.

I refactored the setup of the Autoencoder (PyTorch Version) in order to address the points of #434. The order of layers is now consistent with the Tensorflow implementation.

The order of Neurons->BatchNorm->Activation follows the Standard-Implementation of [ResNet](https://arxiv.org/abs/1512.03385)

> We adopt batch normalization (BN) [16] right after each convolution and before activation, following [16].

The printed Network now looks like:

<Details>

```
AutoEncoder(batch_norm=True, batch_size=32, contamination=0.1,
      device=device(type='cpu'), dropout_rate=0.2, epochs=10,
      hidden_activation='relu', hidden_neurons=[64, 32],
      learning_rate=0.001, loss_fn=MSELoss(), preprocessing=True,
      weight_decay=1e-05)
InnerAutoencoder(
  (activation): ReLU()
  (encoder): Sequential(
    (linear0): Linear(in_features=300, out_features=64, bias=True)
    (batch_norm0): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu0): ReLU()
    (dropout0): Dropout(p=0.2, inplace=False)
    (linear1): Linear(in_features=64, out_features=32, bias=True)
    (batch_norm1): BatchNorm1d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu1): ReLU()
    (dropout1): Dropout(p=0.2, inplace=False)
  )
  (decoder): Sequential(
    (linear0): Linear(in_features=32, out_features=64, bias=True)
    (batch_norm0): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu0): ReLU()
    (dropout0): Dropout(p=0.2, inplace=False)
    (linear1): Linear(in_features=64, out_features=300, bias=True)
    (relu1): ReLU()
  )
)
```

</Details>

### All Submissions Basics:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Have you checked all [Issues](../../issues) to tie the PR to a specific one?

### All Submissions Cores:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?
* [x] Does your submission pass tests, including CircleCI, Travis CI, and AppVeyor?
* [x] Does your submission have appropriate code coverage? The cutoff threshold is 95% by Coversall.

### CODE REVIEW

File: pyod/models/auto_encoder_torch.py [hunk 1]
```diff
@@ -44,52 +44,74 @@ def __getitem__(self, idx):
         return torch.from_numpy(sample), idx
 
 
-class inner_autoencoder(nn.Module):
+class InnerAutoencoder(nn.Module):
     def __init__(self,
                  n_features,
-                 hidden_neurons=[128, 64],
+                 hidden_neurons=(128, 64),
```
Comment (Lucew) [inline pyod/models/auto_encoder_torch.py:50]: Python functions should never have mutable default arguments (as they live in global scope and are only initialized once).

File: pyod/models/auto_encoder_torch.py [hunk 2]
```diff
@@ -44,52 +44,74 @@ def __getitem__(self, idx):
         return torch.from_numpy(sample), idx
 
 
-class inner_autoencoder(nn.Module):
+class InnerAutoencoder(nn.Module):
```
Comment (Lucew) [inline pyod/models/auto_encoder_torch.py:47]: Changed to CamelCase class naming (following PEP8)

**Comment by yzhao062** [comment]
sorry there will be some delay on my end to multiple ddl...will review very soon!! thanks

**Comment by Lucew** [comment]
Yeah, no worries. I know that life happens sometimes. Just let me know, once I can do something to bring this further!

**Comment by yzhao062** [comment]
perfect it is in! I will make some local minor updates as needed :) thanks!!
